[Doc] Delta 4.0 - Delta Connect documentation (#3232)
delta-io/delta: docs/source/delta-spark-connect.md

## Motivation

Delta Connect is expected to bring the same benefits as Spark Connect:

1. Easier upgrading to more recent versions of Spark and <Delta>, as the client interface is completely decoupled from the server.
2. Simpler integration of Spark and <Delta> with developer tooling. IDEs no longer have to integrate with the full Spark and <Delta> implementation, and instead can integrate with a thin-client.
3. Support for languages other than Java/Scala and Python. Clients "merely" have to generate Protocol Buffers and therefore become simpler to implement.
4. Spark and <Delta> will become more stable, as user code is no longer running in the same JVM as Spark's driver.
5. Remote connectivity. Code can run anywhere now, as there is a gRPC layer between the user interface and the driver.

## How to start the Spark Server with Delta

1. Download `spark-4.0.0-preview1-bin-hadoop3.tgz` from [Spark 4.0.0 First Preview](https://archive.apache.org/dist/spark/spark-4.0.0-preview1).

2. Start the Spark Connect server with the <Delta> Connect plugins:

```bash

> sbin/start-connect-server.sh \ 

Author: Does this sounds good @tomvanbussel ?

> sbin/start-connect-server.sh \ 

Reviewer: Sounds like a good idea 👍

  --packages org.apache.spark:spark-connect_2.13:4.0.0-preview1,io.delta:delta-connect-server_2.13:4.0.0-preview1,io.delta:delta-connect-common_2.13:4.0.0-preview1,com.google.protobuf:protobuf-java:3.25.1 \ 
  --conf "spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension" \
  --conf "spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog" \
  --conf "spark.connect.extensions.relation.classes=org.apache.spark.sql.connect.delta.DeltaRelationPlugin" \
  --conf "spark.connect.extensions.command.classes=org.apache.spark.sql.connect.delta.DeltaCommandPlugin"
```

## How to use the Python Spark Connect Client with Delta

The <Delta> Connect Python client is included in the same PyPi package as <Delta> Spark.

1. `pip install pyspark==4.0.0`

2. `pip install delta-spark==4.0.0`

3. There is no difference in usage compared to the classic way. We just need to pass in a remote `SparkSession` (instead of a local one) to the `DeltaTable` API.

An example:

```python
import io.delta.tables._
import org.apache.spark.sql.functions._

val deltaTable = DeltaTable.forName(spark, "my_table")

> deltaTable.update(

Author: We already have `forName` in the example, which represents the preview, the `update` is there to represent the final release. @andreaschat-db

> deltaTable.update(

Reviewer 2: Actually, `forName` and `forPath` do not scan the table right? They just return a `DeltaTable`. To scan the table we also need to add support for `toDF`?

Furthermore, shall we add a comment on top of line 54 that this operation is not supported in preview 1?

> deltaTable.update(

Author: @andreaschat-db Done :)

> deltaTable.update(

Reviewer 2: Great. What about my question above?

> deltaTable.update(

Author: @andreaschat-db I added to the table but sure let's make a concrete example with `toDF` as well.

> deltaTable.update(

Reviewer 2: Great.

  condition = expr("id % 2 == 0"),
  set = Map("id" -> expr("id + 100")))
```

## How to use the Scala Spark Connect Client with Delta

> .. note:: These set of steps are going to be simplified with final Spark 4.0 and Delta 4.0

Reviewer 3: ```suggestion
.. note:: These steps are going to be simplified with final Spark 4.0 and Delta 4.0
```

Make sure you are using Java 17!

> 1. `export SPARK_REMOTE="sc://localhost:15002"`

Reviewer 2: Where does the port number here come from? I remember in some tests using port 15003? I do not see it being passed in the server initialization above.

By the way, we need to test all steps in the documentation.

Author: @andreaschat-db It's from https://spark.apache.org/docs/4.0.0-preview1/spark-connect-overview.html#configure-client-server-connection

4. Start the Spark Connect client version 4.0.0 First Preview:

```bash
cs launch --jvm zulu:17.0.11 --scala 2.13.13 -r \
m2Local com.lihaoyi:::ammonite:3.0.0-M2 org.apache.spark::spark-connect-client-jvm:4.0.0-preview1 io.delta:delta-connect-client_2.13:4.0.0-preview1 io.delta:delta-connect-common_2.13:4.0.0-preview1 com.google.protobuf:protobuf-java:3.25.1 \
--java-opt --add-opens=java.base/java.nio=ALL-UNNAMED \
-M org.apache.spark.sql.application.ConnectRepl
```

An example:
    
```scala
import io.delta.tables._
import org.apache.spark.sql.functions._

val deltaTable = DeltaTable.forName(spark, "my_table")

deltaTable.update(
condition = expr("id % 2 == 0"),
set = Map("id" -> expr("id + 100")))
```

In the future, when [spark-connect-repl](https://spark.apache.org/docs/4.0.0-preview1/spark-connect-overview.html#use-spark-connect-for-interactive-analysis) moves to Spark 4.0 and above, we will use `spark-connect-repl` instead of `cs`.

## Preview Limitations

The feature introduces a limited set of supported operations in <Delta> 4.0 preview and expands it in <Delta> 4.0 and above.

.. csv-table::

> :header: "Supported operations - Delta 4.0 preview", "Supported operations - Delta 4.0 release"

Reviewer 2: Scans are supported in the preview right?

Reviewer 2: I was wondering whether the table here should have a comment to clarify this.

Author: Isn't it `The feature introduces a limited set of supported operations in <Delta> 4.0 preview and expands it in <Delta> 4.0 and above.` above @andreaschat-db ?

Reviewer 2: My question here was whether we should clarify that scans are supported in the preview. I am not sure that everyone can understand that by supporting `forName` and `toDF` we support scans.

Author: @andreaschat-db Sure we can make it a bit clearer by adding an explicit comment here.

---
description: Learn about Delta Connect - Spark Connect in Delta.
---

> # Delta Connect (or Spark Connect in Delta)

Reviewer 3: ```suggestion
# Delta Connect (or Spark Connect Support in Delta)
```

> .. note:: This feature is available in preview in <Delta> 4.0 and above.

Reviewer 3: ```suggestion
.. note:: This feature is available in preview in <Delta> 4.0.
```

> <Delta> Connect adds [Spark Connect](https://spark.apache.org/docs/latest/spark-connect-overview.html) support to Scala and Python APIs of <Delta> Lake for Apache Spark. Spark Connect is a new project released in Apache Spark 4.0 that adds a decoupled client-server infrastructure which allows remote connectivity from Spark from everywhere. <Delta> Connect makes the `DeltaTable` interfaces compatible with the new Spark Connect protocol.

Reviewer 3: ```suggestion
Delta Connect adds [Spark Connect](https://spark.apache.org/docs/latest/spark-connect-overview.html) support to <Delta> Lake for Apache Spark. Spark Connect is a new initiative that adds a decoupled client-server infrastructure which allows remote connectivity from Spark from everywhere. Delta Connect allows all Delta Lake operations to work in your application running as a client connected to the Spark server.
```

## Motivation

> <Delta> Connect is expected to bring the same benefits as Spark Connect:

Reviewer 3: ```suggestion
Delta Connect is expected to bring the same benefits as Spark Connect:
```

1. Easier upgrading to more recent versions of Spark and <Delta>, as the client interface is completely decoupled from the server.
2. Simpler integration of Spark and <Delta> with developer tooling. IDEs no longer have to integrate with the full Spark and <Delta> implementation, and instead can integrate with a thin-client.
3. Support for languages other than Java/Scala and Python. Clients "merely" have to generate Protocol Buffers and therefore become simpler to implement.
4. Spark and <Delta> will become more stable, as user code is no longer running in the same JVM as Spark's driver.
5. Remote connectivity. Code can run anywhere now, as there is a gRPC layer between the user interface and the driver.

> ## Supported Delta Connect operations

Reviewer 3: Move this section to the end and rename it as "Preview Limitations"

Reviewer 3: And make it a shorter list of things that are known to not work

Author: @tdas How can we make it shorter? I already shortened the list (you can do many things with `merge` and `optimize`), shall I omit `createIfNotExists` and combine `restoreToVersion` and `restoreToTimestamp` into `restore`?

Author: @tdas Well for Preview we only support `forName`, and `forPath`. In the official 4.0 release we plan to support all operations in https://github.com/delta-io/delta/blob/master/python/delta/tables.py for Delta Connect.

Does this answer your question? I made a board in the doc.

The feature introduces a limited set of supported operations in <Delta> 4.0 preview and expands it in <Delta> 4.0 and above.

.. csv-table::
:header: "Supported operations - Delta 4.0 preview", "Supported operations - Delta 4.0 release"

"`forPath`","`forPath`"
"`forName`","`forName`"
"`alias`","`alias`"
"`toDF`","`toDF`"
,"`generate`"
,"`vacuum`"
,"`delete`"
,"`update`"
,"`merge`"
,"`history`"
,"`detail`"
,"`convertToDelta`"
,"`create`"
,"`createIfNotExists`"
,"`replace`"
,"`createOrReplace`"
,"`isDeltaTable`"
,"`upgradeTableProtocol`"
,"`restoreToVersion`"
,"`restoreToTimestamp`"
,"`optimize`"
,"`clone`"

> ## How to start the Delta Connect Server

Reviewer 3: ```suggestion
## How to start the Spark Server with Delta
```

1. Download `spark-4.0.0-preview1-bin-hadoop3.tgz` from [Spark 4.0.0 First Preview](https://archive.apache.org/dist/spark/spark-4.0.0-preview1).

2. Start the Spark Connect server with the <Delta> Connect plugins:

```bash
sbin/start-connect-server.sh \ 
  --packages org.apache.spark:spark-connect_2.13:4.0.0-preview1,io.delta:delta-connect-server_2.13:4.0.0-preview1,io.delta:delta-connect-common_2.13:4.0.0-preview1,com.google.protobuf:protobuf-java:3.25.1 \ 
  --conf "spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension" \
  --conf "spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog" \
  --conf "spark.connect.extensions.relation.classes=org.apache.spark.sql.connect.delta.DeltaRelationPlugin" \
  --conf "spark.connect.extensions.command.classes=org.apache.spark.sql.connect.delta.DeltaCommandPlugin"
```

> ## How to use the Delta Connect Python Client

Reviewer 3: ```suggestion
## How to use the Python Spark Connect Client with Delta 
```

The <Delta> Connect Python client is included in the same PyPi package as <Delta> Spark.

1. `pip install pyspark==4.0.0`

2. `pip install delta-spark==4.0.0`

3. There is no difference in usage compared to the classic way. We just need to pass in a remote `SparkSession` (instead of a local one) to the `DeltaTable` API.

An example:

```python
import io.delta.tables._
import org.apache.spark.sql.functions._

val deltaTable = DeltaTable.forName(spark, "my_table")

deltaTable.update(
  condition = expr("id % 2 == 0"),
  set = Map("id" -> expr("id + 100")))
```

> ## How to use the Delta Connect Scala Client

Reviewer 3: ```suggestion
## How to use the Scala Spark Connect Client with Delta 
```

> ## How to use the Scala Spark Connect Client with Delta

Reviewer 3: Add a note, saying these set of steps are going to be simplified with final Spark 4.0 and Delta 4.0

```bash
sbin/start-connect-server.sh \ 
  --packages org.apache.spark:spark-connect_2.13:4.0.0-preview1,io.delta:delta-connect-server_2.13:4.0.0-preview1,io.delta:delta-connect-common_2.13:4.0.0-preview1,com.google.protobuf:protobuf-java:3.25.1 \ 
  --conf "spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension" \
  --conf "spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog" \
  --conf "spark.connect.extensions.relation.classes=org.apache.spark.sql.connect.delta.DeltaRelationPlugin" \
  --conf "spark.connect.extensions.command.classes=org.apache.spark.sql.connect.delta.DeltaCommandPlugin"
```

## How to use the Python Spark Connect Client with Delta

The <Delta> Connect Python client is included in the same PyPi package as <Delta> Spark.

1. `pip install pyspark==4.0.0`

2. `pip install delta-spark==4.0.0`

3. The usage is the same as Spark Connect. We just need to pass in a remote `SparkSession` (instead of a local one) to the `DeltaTable` API.

An example:

```python
import io.delta.tables._
import org.apache.spark.sql.functions._

val deltaTable = DeltaTable.forName(spark, "my_table")

> # Update will only be supported in the final release of Delta 4.0.

Reviewer 2: At a later preview stage of Delta 4.0?

Author: You are right, let's say it is not supported in the first preview @andreaschat-db

```bash
sbin/start-connect-server.sh \ 
  --packages org.apache.spark:spark-connect_2.13:4.0.0-preview1,io.delta:delta-connect-server_2.13:4.0.0-preview1,io.delta:delta-connect-common_2.13:4.0.0-preview1,com.google.protobuf:protobuf-java:3.25.1 \ 
  --conf "spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension" \
  --conf "spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog" \
  --conf "spark.connect.extensions.relation.classes=org.apache.spark.sql.connect.delta.DeltaRelationPlugin" \
  --conf "spark.connect.extensions.command.classes=org.apache.spark.sql.connect.delta.DeltaCommandPlugin"
```

## How to use the Python Spark Connect Client with Delta

The <Delta> Connect Python client is included in the same PyPi package as <Delta> Spark.

1. `pip install pyspark==4.0.0`

2. `pip install delta-spark==4.0.0`

3. The usage is the same as Spark Connect. We just need to pass in a remote `SparkSession` (instead of a local one) to the `DeltaTable` API.

An example:

```python
import io.delta.tables._
import org.apache.spark.sql.functions._

val deltaTable = DeltaTable.forName(spark, "my_table")
deltaTable.toDF().show()

> // Update will not be supported in the first of Delta 4.0.

Reviewer 2: Same for the scala example below.
```suggestion
// Update will not be supported in the first preview of Delta 4.0.
```
By the way, I though python comments start with #?

--java-opt --add-opens=java.base/java.nio=ALL-UNNAMED \
-M org.apache.spark.sql.application.ConnectRepl
```

An example:
    
```scala
import io.delta.tables._
import org.apache.spark.sql.functions._

val deltaTable = DeltaTable.forName(spark, "my_table")
deltaTable.toDF.show()

// Update will not be supported in the first of Delta 4.0.
deltaTable.update(
condition = expr("id % 2 == 0"),
set = Map("id" -> expr("id + 100")))
```

In the future, when [spark-connect-repl](https://spark.apache.org/docs/4.0.0-preview1/spark-connect-overview.html#use-spark-connect-for-interactive-analysis) moves to Spark 4.0 and above, we will use `spark-connect-repl` instead of `cs`.

## Preview Limitations

The feature introduces a limited set of supported operations in <Delta> 4.0 preview and expands it in <Delta> 4.0 and above.

.. csv-table::
:header: "Supported operations - Delta 4.0 preview", "Supported operations - Delta 4.0 release"

"`forPath`","`forPath`"
"`forName`","`forName`"
"`toDF`","`toDF`"

> ,"`alias`"

Reviewer 2: By the way, didn't we also add support for that? I remember seeing this in both the scala and python clients.

Author: @andreaschat-db Once I finish testing this to be sure I will update it later, this is fine as discussed with @tdas
